### PR TITLE
Fix FP `used-before-assignment` for statements guarded under same test

### LINKED ISSUE #8167: False positive  `used-before-assignment` when usage and assignment guarded by the same test in different statements
### Bug description

The `used-before-assignment` check reports false positive for using the same variable inside multiple `if __name__ == '__main__'` statements.

```
# pylint: disable=missing-docstring

if __name__ == '__main__':
    X = 1

if __name__ == '__main__':
    print(X)
```

### Configuration

_No response_

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:7:10: E0601: Using variable 'X' before assignment (used-before-assignment)

-----------------------------------
Your code has been rated at 0.00/10
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 2.16.0
astroid 2.14.1
Python 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
Closes #8167

### CODE REVIEW

File: tests/functional/u/used/used_before_assignment.py [hunk 1]
```diff
@@ -116,3 +116,50 @@ def turn_on2(**kwargs):
...
+if give_me_nothing():
+    print(WE_HAVE_ZERO)
+
+
+# Different tests, different values
+def give_me_none():
+    return None
+
+if give_me_none():
+    print(ALL_DONE)  # [used-before-assignment]
```
Comment (Pierre-Sassoulas) [inline tests/functional/u/used/used_before_assignment.py:165]: I like those tests 👍

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -819,6 +822,38 @@ def _uncertain_nodes_in_false_tests(
 
         return uncertain_nodes
 
+    @staticmethod
+    def _node_guarded_by_same_test(node: nodes.NodeNG, other_if: nodes.If) -> bool:
+        """Identify if `node` is guarded by an equivalent test as `other_if`.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:827]: There's a lot of inference here, but the primer time vary in the same range than it usually does, so the performance impact is probably acceptable.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, the primer result bring joy to my heart.

File: pylint/checkers/variables.py (hunk 2 above)
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:827]: Yeah, I'm assuming that all of these nodes are inferred anyway elsewhere in the variables checker, so we're hitting the cache.

**Comment by Pierre-Sassoulas** [comment]
I think this is going very hard to backport (I tried but there's a lot of concurrent changes), what do you think @jacobtylerwalls ?

**Comment by jacobtylerwalls** [comment]
I tried also. I'm okay waiting for the 3.0 beta for this. music21 can always use bleeding-edge pylint 🗡️

**Pull request merged**